### PR TITLE
Jira #1472 added new actions, refactored others, deleted one

### PR DESCRIPTION
**Pull request by feyzullah**
Added actions;
*add-watcher-to-issue
*assign-issue
Refactored actions:
*add-attachment-to-issue
*add-comment-to-issue
*create-issue
*create-version
*delete-project
*get-all-projects
*get-issue
*get-task
*get-transitions
*get-user
*list-issue-comments
*make-api-call
*transition-issue
*update-comment
*update-issue
Deleted action(Deleted because it is fed with webhook events and currently we have no webhook integration):
*transit-status-for-linked-issues

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

File: components/jira/actions/add-comment-to-issue/add-comment-to-issue.mjs [hunk 1]
```diff
@@ -1,97 +1,71 @@
...
       description: "Extra properties of any type may be provided to this object.",
       optional: true,
     },
+    expand: {
+      type: "string",
+      label: "Expand",
+      description: "The Jira REST API uses resource expansion, which means that some parts of a resource are not returned unless specified in the request. This parameter accepts `renderedBody`, which returns the comment body rendered in HTML.",
+      optional: true,
+    },
   },
```
Comment (michelle0927) [inline components/jira/actions/add-comment-to-issue/add-comment-to-issue.mjs:46]: `properties`, `additionalProperties`, and `expand` are used in multiple components, so they should be moved to the app file. Then you'd use propDefinitions for them just like `issueIdOrKey`. You can overwrite the definitions in the components if they need to be different.

This applies to all props that are used in multiple components.

File: components/jira/actions/create-issue/create-issue.mjs [hunk 2]
```diff
@@ -1,228 +1,213 @@
...
-      optional: true,
+      label: "Reporter ID",
+      description: "Reporter ID of the issue.",
     },
     labels: {
       type: "any",
+      label: "Labels",
       description: "Labels associated to the issue, such as \"bugfix\", \"blitz_test\".",
       optional: true,
     },
```
Comment (michelle0927) [inline components/jira/actions/create-issue/create-issue.mjs:58]: We should use async options to get the available `labels` for the user to select from, and move this prop to the app file.
https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-labels/#api-rest-api-3-label-get

File: components/jira/actions/create-issue/create-issue.mjs [hunk 3]
```diff
@@ -1,228 +1,213 @@
...
-    assignee_id: {
+    assigneeId: {
       type: "string",
-      description: "User assigned to fixing the issue.",
+      label: "Assignee ID",
+      description: "ID of the user assigned to the issue.",
       optional: true,
     },
```
Comment (michelle0927) [inline components/jira/actions/create-issue/create-issue.mjs:70]: We should use async options to get the available users to select from, and move this prop to the app file.
https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-users/#api-rest-api-3-users-search-get

File: components/jira/actions/list-issue-comments/list-issue-comments.mjs [hunk 4]
```diff
@@ -1,33 +1,34 @@
...
+        jira,
+        "issueIdOrKey",
+      ],
     },
     startAt: {
       type: "integer",
+      label: "Start at",
       description: "The index of the first item to return in a page of results (page offset).\nDefault: `0`, Format: `int64`.",
       optional: true,
     },
```
Comment (michelle0927) [inline components/jira/actions/list-issue-comments/list-issue-comments.mjs:22]: This prop `startAt` can be removed. We should handle any pagination ourselves either in the component or in the app file. I like to handle pagination in the app file, and then calling `this.jira.listIssueComments` would return a list of paginated results.

File: components/jira/jira.app.mjs [hunk 5]
```diff
@@ -1,11 +1,257 @@
...
+  propDefinitions: {
+    projectID: {
+      type: "string",
+      label: "Project ID",
+      description: "The project ID.",
+      async options() {
+        const resp = await this._makeRequest({
+          method: "GET",
+          path: "/project",
+        });
```
Comment (michelle0927) [inline components/jira/jira.app.mjs:15]: We should use utilize the method `getAllProjects` here instead of `_makeRequest`. Also make sure to handle pagination in case there are multiple pages of projects.

File: components/jira/jira.app.mjs [hunk 6]
```diff
@@ -1,11 +1,257 @@
...
+          label: e.name,
+          value: e.id,
+        }));
+      },
+    },
+    issueIdOrKey: {
+      type: "string",
+      label: "Issue id or key",
+      description: "The ID or key of the issue where the attachment will be added to.",
+    },
```
Comment (michelle0927) [inline components/jira/jira.app.mjs:45]: Can we use async options for `issueIdOrKey` to list the available issues?
https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-issue-search/#api-rest-api-3-search-get

File: components/jira/jira.app.mjs [hunk 7]
```diff
@@ -1,11 +1,257 @@
...
+    accountId: {
+      type: "string",
+      label: "Account Id",
+      description: "The account ID of the user, which uniquely identifies the user across all Atlassian products, For example, `5b10ac8d82e05b22cc7d4ef5`, ",
+    },
```
Comment (michelle0927) [inline components/jira/jira.app.mjs:50]: I don't think having a prop for `accountId` is necessary. We can retrieve the authenticated user's accountId from the API:
https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-myself/#api-rest-api-3-myself-get

File: components/jira/jira.app.mjs [hunk 8]
```diff
@@ -1,11 +1,257 @@
...
+        ...args,
+      });
+    },
+    async getAllProjects({ ...args } = {}) {
+      return await this._makeRequest({
+        method: "GET",
+        path: "/project",
+        ...args,
+      });
+    },
```
Comment (michelle0927) [inline components/jira/jira.app.mjs:175]: This endpoint is actually deprecated. I think we should use `/project/search`
https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-projects/#api-rest-api-3-project-search-get
And make sure to handle pagination
https://developer.atlassian.com/cloud/jira/platform/rest/v3/intro/#pagination

File: components/jira/jira.app.mjs [hunk 9]
```diff
@@ -1,11 +1,257 @@
...
+      return await this._makeRequest({
+        method: "GET",
+        path: "/user",
+        ...args,
+      });
+    },
+    async listIssueComments({
+      issueIdOrKey,
+      ...args
+    } = {}) {
```
Comment (michelle0927) [inline components/jira/jira.app.mjs:218]: `listIssueComments` should handle pagination in case there are multiple pages of comments.

**Review by michelle0927 — CHANGES_REQUESTED** [review]
Thanks for updating all these. I made some comments about adding `async options` to some props, but if there are any other props that we can add `async options` to, we should do so as well.
Also, can you add a package.json file?

File: components/jira/actions/add-comment-to-issue/add-comment-to-issue.mjs (hunk 1 above)
Comment (feyzullah) [reply components/jira/actions/add-comment-to-issue/add-comment-to-issue.mjs:46]: I will do in this way, thanks!

File: components/jira/actions/create-issue/create-issue.mjs (hunk 2 above)
Comment (feyzullah) [reply components/jira/actions/create-issue/create-issue.mjs:58]: Actually I did not realize it, good catch thanks!

File: components/jira/actions/create-issue/create-issue.mjs (hunk 3 above)
Comment (feyzullah) [reply components/jira/actions/create-issue/create-issue.mjs:70]: I will do it in this way too

File: components/jira/actions/list-issue-comments/list-issue-comments.mjs (hunk 4 above)
Comment (feyzullah) [reply components/jira/actions/list-issue-comments/list-issue-comments.mjs:22]: makes sense thank you

File: components/jira/jira.app.mjs (hunk 6 above)
Comment (feyzullah) [reply components/jira/jira.app.mjs:45]: I think there can be a lot of issues and user cannot find by selecting from list and loading more and more. And when fetching them, they come with some extra data, which can make it huge. This usage may not be very handful. What do you think? Please correct me if I am wrong

File: components/jira/jira.app.mjs (hunk 7 above)
Comment (feyzullah) [reply components/jira/jira.app.mjs:50]: This one is actually used for not users themselves but other users who can be assigned to an issue. For example, user A assigns an issue to user B, and this accountId is used for user B in this case. How about a paginated option prop for users here?

File: components/jira/jira.app.mjs (hunk 8 above)
Comment (feyzullah) [reply components/jira/jira.app.mjs:175]: Sorry for that, I guess I missed it

File: components/jira/jira.app.mjs (hunk 9 above)
Comment (feyzullah) [reply components/jira/jira.app.mjs:218]: Doing in this way, thanks!

File: components/jira/jira.app.mjs (hunk 5 above)
Comment (feyzullah) [reply components/jira/jira.app.mjs:15]: Doing in this way, thanks!

**Comment by dannyroosevelt** [comment]
@feyzullah FYI we've added to additional scopes you requested -- let me know if you are unblocked.

**Comment by feyzullah** [comment]
@dannyroosevelt we are having some scope problems again. I tried almost everything and every combination to get some of the events from webhooks, but I guess I was searching it in wrong places. What I found out is that we again will need some other auth scopes. I hope this will be the last one, but I don't trust Jira docs also.
I guess we have these scopes currently;
* `manage:jira-project`
* `manage:jira-configuration`
* `manage:jira-data-provider`
* `offline_access`
* `write:jira-work`
* `read:me`
* `read:jira-user`
* `write:webhook:jira`
* `read:field:jira`
* `read:project:jira`
* `delete:webhook:jira`

According to this [document](https://developer.atlassian.com/cloud/jira/platform/webhooks/#using-the-rest-api--registration);

For `jira:issue_created`, `jira:issue_updated` and `jira:issue_deleted` it is said that the following scope is needed(However I can get these events correctly);
* `read:issue-details:jira`

For `comment_created`, `comment_updated` and `comment_deleted` the following scopes are needed(I could not make it work);
* `read:comment.property:jira`
* `read:comment:jira`
* `read:epic:jira-software`
* `read:group:jira`
* `read:issue.property:jira`
* `read:issue-type:jira`
* `read:project:jira`
* `read:project-role:jira`
* `read:status:jira`
* `read:user:jira`

And for `issue_property_set` and `issue_property_deleted` these scopes are needed(I could not make it work too);
* `read:epic:jira-software`
* `read:issue.property:jira`
* `read:issue-type:jira`
* `read:project:jira`
* `read:status:jira`
* `read:user:jira`

In short, when already existing scopes are eliminated, the following scopes are left;
* `read:issue-details:jira`
* `read:comment.property:jira`
* `read:comment:jira`
* `read:epic:jira-software`
* `read:group:jira`
* `read:issue.property:jira`
* `read:issue-type:jira`
* `read:project-role:jira`
* `read:status:jira`
* `read:user:jira`

Can you please add these and let me try again?

**Comment by dannyroosevelt** [comment]
> 

So to confirm, we need to add these scopes?
```
read:issue-details:jira
read:comment.property:jira
read:comment:jira
read:epic:jira-software
read:group:jira
read:issue.property:jira
read:issue-type:jira
read:project-role:jira
read:status:jira
read:user:jira
```

**Comment by ctrlaltdylan** [comment]
@dannyroosevelt @feyzullah Have the scoping issues been addressed?

We have a customer expericing a bug on the **Create Ticket** action where the `assignee` is missing,  I wonder if this update contains the fix.

Screenshot 2022-06-06 at 13 53 58

**Comment by feyzullah** [comment]
> 

Hi @ctrlaltdylan we decided not to add all of the scopes. They are actually used for some event sources. A mixed fix is coming for some new actions/sources and refactors

**Comment by feyzullah** [comment]
Because another commit has been done before I merge this, I needed to refactor again.

**Comment by feyzullah** [comment]
> 

@ctrlaltdylan by the way if **Create Ticket** action is another name for **jira_create_issue**, yes it is going to be fixed

**Comment by feyzullah** [comment]
> 

@ctrlaltdylan actually I finished all other job except refactoring last commit. I was planning to commit all before but I could not. Now I am planning to commit today. I will also ping the reviewer to speed up the process
